unixfs/src/file/adder.rs

```diff
@@ -0,0 +1,699 @@
+use cid::Cid;
+
+use crate::pb::{FlatUnixFs, PBLink, UnixFs, UnixFsType};
+use quick_protobuf::{MessageWrite, Writer};
+use std::borrow::Cow;
+use std::fmt;
+
+use sha2::{Digest, Sha256};
+
+/// File tree builder. Implements `Default` which tracks the recent defaults.
+///
+/// Custom file tree builder can be created with [`FileAdder::builder()`] and configuring the
+/// chunker and collector.
+///
+/// Current implementation maintains an internal buffer for the block creation and uses a
+/// non-customizable hash function to produce Cid version 0 links. Currently does not support
+/// inline links.
+#[derive(Default)]
+pub struct FileAdder {
+    chunker: Chunker,
+    collector: Collector,
+    block_buffer: Vec<u8>,
+    // all unflushed links as a flat vec; this is compacted as we grow and need to create a link
+    // block for the last N blocks, as decided by the collector
+    unflushed_links: Vec<(usize, Cid, u64, u64)>,
```

Comment:
the `(usize, Cid, u64, u64)` type appears 5 times and is pretty long; I'd use a `type` alias for it for extra readability

Reply:
Done, agreed. It's just so much nicer to write these tuples everywhere and have to look up which field was which ... Something I need to unlearn :)